Migrate HFDeepSpeedConfig from trfrs to accelerate

pacman100 opened the pull request:
### What does this PR do?
1. Migrates HFDeepSpeedConfig from transformers repo to accelerate repo as it is generic enough and specific bits remain in transformers.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

pacman100 committed:
Migrate HFDeepSpeedConfig from trfrs to accelerate

pacman100 committed:
add `accelerate` to testing dep

sgugger commented on src/transformers/deepspeed.py:
```diff
@@ -283,36 +160,28 @@ def trainer_config_finalize(self, args, model, num_training_steps):
-        return _hf_deepspeed_config_weak_ref().is_zero3()
-    else:
-        return False
+    accelerator = Accelerator()
+    if hasattr(accelerator.state, "deepspeed_plugin"):
+        return (
+            accelerator.state.deepspeed_plugin.hf_ds_config.is_zero3()
+            and accelerator.state.deepspeed_plugin.zero3_init_flag
+        )
```
The `Accelerator` object should not be created behind the user back as they might set their own options in it. We should instantiate `AcceleratorState` here (which is always the same) and use it directly.

sgugger commented on src/transformers/deepspeed.py:
```diff
@@ -283,36 +160,28 @@ def trainer_config_finalize(self, args, model, num_training_steps):
-    else:
-        return None
+    ds_config = None
+    accelerator = Accelerator()
+    if hasattr(accelerator.state, "deepspeed_plugin"):
```
Same here.

sgugger commented on src/transformers/training_args.py:
```diff
@@ -1097,13 +1098,22 @@ def __post_init__(self):
             self.hf_deepspeed_config = HfTrainerDeepSpeedConfig(self.deepspeed)
             self.hf_deepspeed_config.trainer_config_process(self)
 
+            from accelerate import Accelerator
+            from accelerate.utils import DeepSpeedPlugin
+
+            _ = Accelerator(
+                deepspeed_plugin=DeepSpeedPlugin(hf_ds_config=self.hf_deepspeed_config, zero3_init_flag=True)
+            )
```
Same here, we shouldn't define an `Accelerator` here. Since all `AcceleratorState` share the same inner dict, setting its deepspeed plugin directly will work normally.

pacman100 replied:
In `training_args.py` when user is running trainer scripts, Accelerator object is nowhere created and hence if we only have to create `AcceleratorState` then we have to pass `_from_accelerator=True` to it. Is that fine?

sgugger replied:
It's not ideal, as that creation should be done by the user only at the right point in their scripts. However it's unlikely a user using the Trainer will re-use the Accelerator object so it can work if there is really no alternative.

sgugger commented on tests/deepspeed/test_deepspeed.py:
```diff
@@ -165,7 +167,9 @@ def tearDown(self):
 
         # reset the ds config global so that tests state doesn't leak
-        unset_hf_deepspeed_config()
+        with mockenv_context(**self.dist_env_1_gpu):
+            accelerator = Accelerator()
+            accelerator.state.initialized = False
```
I don't think this changes the fast `accelerator.state` keeps the deepspeed config here. You should change `AcceleratorState` directly.

sgugger commented on tests/deepspeed/test_deepspeed.py:
```diff
@@ -237,9 +241,9 @@ def setUp(self):
-
-        # reset the ds config global so that tests state doesn't leak
-        unset_hf_deepspeed_config()
+        with mockenv_context(**self.dist_env_1_gpu):
+            accelerator = Accelerator()
+            accelerator.state.initialized = False
 
```
Same here.

sgugger commented on tests/deepspeed/test_deepspeed.py:
```diff
@@ -728,6 +732,9 @@ def test_config_object(self):
             self.assertTrue(is_deepspeed_zero3_enabled())
 
             # test zero3 is disabled
+            with mockenv_context(**self.dist_env_1_gpu):
+                accelerator = Accelerator()
+                accelerator.state.initialized = False
```
And there.

sgugger commented on tests/deepspeed/test_deepspeed.py:
```diff
@@ -736,7 +743,9 @@ def test_config_object(self):
             del trainer
-            # now weakref should gc the global and we shouldn't get anything here
+            with mockenv_context(**self.dist_env_1_gpu):
+                accelerator = Accelerator()
+                del accelerator.state.deepspeed_plugin
```
Del directly on AcceleratorState here.

sgugger reviewed (commented):
Thanks for working on this! Lots of tests/change are done on `accelerator.state` when they should be done on an instance of `AcceleratorState` instead (all isntances of that class have the same inner dict).

pacman100 committed:
addressing comments

pacman100 committed:
addressing comments

Using `_shared_state` and avoiding object creation. This is necessary as `notebook_launcher` in `launcers.py` checks `len(AcceleratorState._shared_state)>0` to throw an error.

sgugger commented on src/transformers/deepspeed.py:
```diff
@@ -283,36 +160,30 @@ def trainer_config_finalize(self, args, model, num_training_steps):
-    else:
-        return False
+    if (
+        "deepspeed_plugin" in AcceleratorState._shared_state
+        and AcceleratorState._shared_state["deepspeed_plugin"] is not None
```
Let's instantiate `AcceleratorState()` instead of looking at its `_sharde_state`.

pacman100 replied:
Hello, at first I was creating `AcceleratorState` object like you recommend in the following comments. But, then I found out that in `notebook_launcher` in `launchers.py`, `len(Accelerator._shared_state)>0` is done to throw error. As much, it will always throw error in `notebook_launcher` as calling `AutoModel.from_pretrained` will instantiate `AcceleratorState` object in the purposed changes. Also, we need to be careful with calling  `AcceleratorState()` as it was creating and intializing cuda in multiprocessing setup which would eventually throw `cuda already initialized` errors. For all these reasons, I changes from working with state object to its `_shared_state`.

stas00 replied:
why not just copy the transformers implementation verbatim and wrap around it for accelerate needs? The transformers implementation requires no complicated objects with side-effects. Let's keep simple things simple and difficult things possible.

Also heads up, we are discussing having Deepspeed provide this functionality in its core - it may take some weeks months to happens, but it's good to plan to have this whole object management in the deepspeed side. The idea is to have the config object to be part of the deepspeed object, but not require its initialize to be able to use it. There is no API yet, but it'd be something similar to what transformer has except built into deepspeed's API.

pacman100 replied:
Hello Stas, earlier `weak_ref` was needed for checking if stage_3 is enabled or not. The effort of this PR is to get rid of `weak_ref` and `HFDeepSpeedConfig` has been copied verbatim in Accelerate. 

These changes are required to know the deepspeed stage and config without using weakref. 

It's not complicated as such as we are checking Accelerator state's content which will be same/available throughout and thereby overcoming the need for weakref.

stas00 replied:
yes, of course, but if possible, let's keep the API simple for non-accelerator uses. When using with (1) HF Trainer or (2) w/o any trainer at all, there shouldn't be a need to create an Accelerator object since in both of these cases Accelerator is not used. So ideally there should be simple low-level API that lives in the `accelerate` codebase, but doesn't need to really rely on `accelerate` high-end functionality. I hope it makes sense. But I trust that between you and Sylvain you guys will sort it out as I don't know anything about `accelerate`.

sgugger commented on src/transformers/training_args.py:
```diff
@@ -1097,13 +1098,22 @@ def __post_init__(self):
+            from accelerate.state import AcceleratorState
+            from accelerate.utils import DeepSpeedPlugin
+
+            AcceleratorState._shared_state["deepspeed_plugin"] = DeepSpeedPlugin(
```
```suggestion
            AcceleratorState()["deepspeed_plugin"] = DeepSpeedPlugin(
```

sgugger commented on tests/deepspeed/test_deepspeed.py:
```diff
@@ -176,27 +176,30 @@ def test_init_zero3_fp16(self):
         with LoggingLevel(logging.INFO):
             with mockenv_context(**self.dist_env_1_gpu):
+                dschf = HfTrainerDeepSpeedConfig(ds_config)
+                AcceleratorState._shared_state["deepspeed_plugin"] = DeepSpeedPlugin(
```
```suggestion
                AcceleratorState()["deepspeed_plugin"] = DeepSpeedPlugin(
```

sgugger commented on tests/deepspeed/test_deepspeed.py:
```diff
@@ -176,27 +176,30 @@ def test_init_zero3_fp16(self):
         with LoggingLevel(logging.INFO):
             with mockenv_context(**self.dist_env_1_gpu):
+                dschf = HfTrainerDeepSpeedConfig(ds_config)
+                AcceleratorState._shared_state["deepspeed_plugin"] = DeepSpeedPlugin(
```
```suggestion
                AcceleratorState()["deepspeed_plugin"] = DeepSpeedPlugin(
```

sgugger commented on src/transformers/deepspeed.py:
```diff
@@ -283,36 +160,30 @@ def trainer_config_finalize(self, args, model, num_training_steps):
+        "deepspeed_plugin" in AcceleratorState._shared_state
+        and AcceleratorState._shared_state["deepspeed_plugin"] is not None
+    ):
+        ds_config = deepcopy(AcceleratorState._shared_state["deepspeed_plugin"].hf_ds_config.config)
+        if ds_config["gradient_accumulation_steps"] == "auto":
+            ds_config["gradient_accumulation_steps"] = 1
+        if "train_micro_batch_size_per_gpu" not in ds_config or ds_config["train_micro_batch_size_per_gpu"] == "auto":
+            ds_config["train_micro_batch_size_per_gpu"] = 1
+        if ds_config["train_batch_size"] == "auto":
+            del ds_config["train_batch_size"]
```
Why are we removing `autos` here? They could be resolved with the `TrainingArguments` no? Or ar those auto values that Accelerate resolves but not Transformers?

pacman100 replied:
In `Accelerate` autos are handled in `prepare` method call but we will be calling `AutoModel.from_pretrained()` or `AutoModel.from_config()` way before it and as such few compulsory fields will be `auto` in accelerate setup when checking zero init is required in stage 3. So, this handles those situations in accelerate and is needed else deepspeed will throw errors when these fields are not correctly populated. 

Transformers Trainer has no issues as all the important population of `auto` fields already happen in `training_args.py`.

This is needed for accelerate integration to work properly when we want to use zero init in stage 3 for directly sharding parameters to GPUs.

sgugger reviewed (commented):
Looking better already, thanks a lot! I have a couple more suggestions.

pacman100 committed:
resolving comments

1. Use simple API from accelerate to manage the deepspeed config integration
2. Update the related documentation

sgugger commented on docs/source/en/main_classes/deepspeed.mdx:
````diff
@@ -1853,52 +1853,47 @@ In this case you usually need to raise the value of `initial_scale_power`. Setti
 
 ```python
-from transformers.deepspeed import HfDeepSpeedConfig
+from accelerate.state import AcceleratorState
````
Let's put `AcceleratorState` at the top level if we need to use it, so it's just
```suggestion
from accelerate import AcceleratorState
```

sgugger commented on docs/source/en/main_classes/deepspeed.mdx:
```diff
@@ -1853,52 +1853,47 @@ In this case you usually need to raise the value of `initial_scale_power`. Setti
 ds_config = {...}  # deepspeed config object or path to the file
 # must run before instantiating the model to detect zero 3
-dschf = HfDeepSpeedConfig(ds_config)  # keep this object alive
+AcceleratorState.set_deepspeed_config(ds_config)
```
Nice new doc!

sgugger reviewed (approved):
Way way better, I really like it like this.
@stas00 I think this is ready for your review.

stas00 commented on tests/deepspeed/test_deepspeed.py:
```diff
@@ -176,27 +175,26 @@ def test_init_zero3_fp16(self):
         self.assertIn("Detected DeepSpeed ZeRO-3", cl.out)
 
-        # now remove zero optimization
-        del ds_config["zero_optimization"]
-        dschf = HfDeepSpeedConfig(ds_config)
-
-        self.assertFalse(dschf.is_zero3())
-        self.assertFalse(is_deepspeed_zero3_enabled())
```
please revert - stage=0 is a thing (used by new bf16 optim)

stas00 replied:
I think the only thing this PR needs to touch is deepspeed.py to create wrappers to the code that moved, everything else should remain as it was before to keep things BC.

stas00 reviewed (changes requested):
This proposal is not backward compatible.

Please leave everything as it was before on the transformers side, API-wise.

The fact that part of the code moved to Accelerate should be transparent to the user, so please use wrappers to hide that.

As mentioned earlier in a few weeks or months this is going to change again once deepspeed supports this internally, and the code will be adapted to switch to that deepspeed internal API.

Here is an alternative proposal:

leave the transformers code as it is and just copy the code to accelerate and adapt it to your liking. The part of the code you are copying hasn't been touched in a long long time, so it's pretty stable and unlikely to change any time soon.

sgugger commented:
My apologies @stas00 I was (wrongly) thinking the DeepSpeed API was still marked as experimental and therefore was okay with some small breaking changes, but it's not so we shouldn't break things.

So let's limit the changes to the move of the config inside Accelerate, with the rest of the code the exact same indeed.

stas00 commented:
That sounds great, Sylvain. Thank you!

pacman100 committed:
reverting changes and addressing comments

pacman100 committed:
docstring correction

sgugger commented on src/transformers/deepspeed.py:
```diff
@@ -36,128 +36,24 @@ def is_deepspeed_available():
-    the DeepSpeed configuration is not modified in any way.
-
-    Args:
+    This object contains a DeepSpeed configuration dictionary and can be quickly queried for things like zero stage. A
+    `weakref` of this object is stored in the module's globals to be able to access the config from areas where things
+    like the Trainer object is not available (e.g. `from_pretrained` and `_get_resized_embeddings`). Therefore it's
+    important that this object remains alive while the program is still running. [`Trainer`] uses the
+    `HfTrainerDeepSpeedConfig` subclass instead. That subclass has logic to sync the configuration with values of
+    [`TrainingArguments`] by replacing special placeholder values: `"auto"`. Without this special logic the DeepSpeed
+    configuration is not modified in any way.
```
Small nit, we should leave this in paragraphs as before.

sgugger commented on src/transformers/deepspeed.py:
```diff
@@ -36,128 +36,24 @@ def is_deepspeed_available():
     """
 
     def __init__(self, config_file_or_dict):
-        # set global weakref object
```
That comment should be left

sgugger reviewed (approved):
Thanks for amending your PR!

pacman100 committed:
addressing nits

pacman100 committed:
addressing nits

pacman100 committed:
addressing nits 3

stas00 commented on setup.py:
```diff
@@ -257,7 +257,7 @@ def run(self):
 
 extras["sagemaker"] = deps_list("sagemaker")
-extras["deepspeed"] = deps_list("deepspeed")
+extras["deepspeed"] = deps_list("deepspeed", "accelerate")
```
and will need to bumpup `"accelerate>=0.9.0",` on line 100 once the accelerate releases a new version before this can be merged.

pacman100 committed:
bumping up the accelerate version to 0.10.0

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_17623). All of your documentation changes will be reflected on that endpoint.

pacman100 committed:
resolving import

pacman100 committed:
update setup.py to include deepspeed dependencies

pacman100 committed:
Merge branch 'main' into smangrul/migrate-HFDeepSpeedConfig

pacman100 committed:
Update dependency_versions_table.py

pacman100 committed:
fixing imports

pacman100 committed:
reverting changes to CI dependencies for "run_tests_pipelines_tf*" tests

These changes didn't help with resolving the failures and I believe this needs to be addressed in another PR.

pacman100 committed:
removing `accelerate` as hard dependency

Resolves issues related to CI Tests

pacman100 committed:
adding `accelerate` as dependency for building docs

resolves failure in Build PR Documentation test

pacman100 committed:
adding `accelerate` as dependency in "dev" to resolve doc build issue

sgugger commented on setup.py:
```diff
@@ -362,6 +363,7 @@ def run(self):
     + extras["docs_specific"]
     + extras["sklearn"]
     + extras["modelcreation"]
+    + extras["accelerate"]
```
Can we put it in `"all"` instead? It should be there and then will be automatically included in dev.

sgugger commented on src/transformers/deepspeed.py:
```diff
@@ -36,7 +34,13 @@ def is_deepspeed_available():
 
 
-class HfDeepSpeedConfig:
+if is_deepspeed_available():
```
The test should probably include `Accelerate` also
```suggestion
if is_accelerate_available() and is_deepspeed_available():
```

pacman100 committed:
resolving comments

1. adding `accelerate` to extras["all"]
2. Including check for accelerate too before import HFDeepSpeedConfig from there

Co-Authored-By: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

sgugger commented:
Thanks for all the work on this, good to merge on my side!

stas00 commented on src/transformers/deepspeed.py:
```diff
@@ -36,7 +34,13 @@ def is_deepspeed_available():
 
 
-class HfDeepSpeedConfig:
+if is_accelerate_available() and is_deepspeed_available():
+    from accelerate.utils.deepspeed import HfDeepSpeedConfig as DeepSpeedConfig
+else:
+    from builtins import object as DeepSpeedConfig
```
Apologies if this has already been discussed, shouldn't it assert if the conditions on line 37 fail?

`HfDeepSpeedConfig` inheriting from `object` will be a bogus class.

pacman100 replied:
Hello @stas, 

The thing is that this for any call of `from_pretrained`, it calls `from .deepspeed import is_deepspeed_zero_stage3`. Now, this call will try import from accelerate and as such fail. If we assert, it will fail again for all calls of `from_pretrained` or `from_config` even if we don't use deepspeed.

And having `accelerate` as hard dependency was causing a lot of Tensorflow tests to fail because installing accelerate will install torch and hence torch tests were being incorrectly run for Tensorflow tests.

When `deepspeed` is being used, assert check is made in `training_args.py` as well in `transformers.deepspeed.HFDeepSpeedConfig` init functions. 

So, when using deepspeed necessary checks are done and everything works as expected.

When not using deepspeed, it is fine if HFDeepSpeedConfig inherits `object` as it will never be used.

stas00 replied:
I thought that there must have been a non-obvious reason for this - thank you for explaining that in details, @pacman100 

Would you kindly document that on line 40, as for a reader that code is bogus w/o the notes. Basically, something like:

```
# will inherit from a dummy `object` if accelerate is not available, so that python succeeds import this file
# deepspeed glue code will not ever use that dummy object as it checks if accelerate is available
```

Much appreciated.

pacman100 committed:
resolving comments
